src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,192 @@
+//! Paths and Unix shells
+//!
+//! MacOS, Linux, FreeBSD, and many other OS model their design on Unix,
+//! so handling them is relatively consistent. But only relatively.
+//! POSIX postdates Unix by 20 years, and each "Unix-like" shell develops
+//! unique quirks over time.
+//!
+//!
+//! Windowing Managers, Desktop Environments, GUI Terminals, and PATHs
+//!
+//! Duplicating paths in PATH can cause performance issues when the OS searches
+//! the same place multiple times. Traditionally, Unix configurations have
+//! resolved this by setting up PATHs in the shell's login profile.
+//!
+//! This has its own issues. Login profiles are only intended to run once, but
+//! changing the PATH is common enough that people may run it twice. Desktop
+//! environments often choose to NOT start login shells in GUI terminals. Thus,
+//! a trend has emerged to place PATH updates in other run-commands (rc) files,
+//! leaving Rustup with few assumptions to build on for fulfilling its promise
+//! to set up PATH appropriately.
+//!
+//! Rustup addresses this by:
+//! 1) using a shell script that updates PATH if the path is not in PATH
+//! 2) sourcing this script in any known and appropriate rc file
+
+use super::canonical_cargo_home;
+use super::*;
+use crate::process;
+use std::path::PathBuf;
+
+pub type Shell = Box<dyn UnixShell>;
+
+#[derive(Debug, PartialEq)]
+pub struct ShellScript {
+    content: &'static str,
+    name: &'static str,
+}
+
+impl ShellScript {
+    pub fn write(&self) -> Result<()> {
+        let cargo_bin = format!("{}/bin", canonical_cargo_home()?);
+        let env_name = utils::cargo_home()?.join(self.name);
+        let env_file = self.content.replace("{cargo_bin}", &cargo_bin);
+        utils::write_file(self.name, &env_name, &env_file)?;
+        Ok(())
+    }
+}
+
+#[allow(dead_code)] // For some reason.
+const POSIX_ENV: &str = include_str!("env");
```

Comment:
Given the size of the template, I'd personally prefer to haave it inline here as a raw string. @kinnison what do you think?

Reply:
I would write that change without complaint but I must Note Ominously:
That will get messy the moment `env.csh`, `env.fish`, or something similar gets implemented.

Reply:
mmm, I would then put those inline in the trait impl for those shells, since we do have that nice structure :).

Reply:
> having it as an external script means we could shellcheck it
> So please ensure it's called something like env.sh so we can do that

-Kinnison, Discord @ 2020-07-13 16:56 UTC

I found his reasoning persuasive here so I'm going to fix that up as well.

Reply:
Or not.